### PR TITLE
[ICC-168] Make validation errors more readable

### PR DESCRIPTION
**Pull request by Garma00**
#### List of Changes

* modified the `getContextPath`
 
#### Motivation and Context

In order to get a more readable output from this function i have removed `io-ts` indexes, here is an example of what we get before and after my changes: 

```bash
[
  'value [1] at [root.1.1.third_party_data.0.id] is not a valid [non empty string]',
  'value [undefined] at [root.1.1.legal_data.0.sender_mail_from] is not a valid [non empty string]',
  'value [undefined] at [root.1.1.legal_data.0.message_unique_id] is not a valid [non empty string]'
]
```

```bash
[
  'value [1] at [root.third_party_data.id] is not a valid [non empty string]',
  'value [undefined] at [root.legal_data.sender_mail_from] is not a valid [non empty string]',
  'value [undefined] at [root.legal_data.message_unique_id] is not a valid [non empty string]'
]
```
 
#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by gunzip** [comment]
Actually, I'd like the current output more since you still get `value [1] at` which makes a little bit sense if `1` is not in the path anymore. Moreover the modified output  (`root.third_party_data.id`) is ambiguos while `root.1.1.third_party_data.id` is not)

File: src/__tests__/responses.test.ts [hunk 1]
```diff
@@ -23,3 +26,37 @@ describe("ResponseSuccessJson", () => {
     expect(mockResponse.json).toHaveBeenCalledWith(kindlessData);
   });
 });
+
+describe("ResponseErrorForbiddenNotAuthorized", () => {
```
Comment (gunzip) [inline src/__tests__/responses.test.ts:30]: why there are these tests in this PR? (it looks unrelated)

File: src/responses.ts [hunk 2]
```diff
@@ -449,16 +456,23 @@ export type IResponseErrorForbiddenNoAuthorizationGroups = IResponse<
 >;
 
 /**
- * The user is not part of any valid authorization groups.
+ * Return an IResponseErrorForbiddenNoAuthorizationGroups with a default detail if not provided
  */
-export const ResponseErrorForbiddenNoAuthorizationGroups: IResponseErrorForbiddenNoAuthorizationGroups = {
+export const getResponseErrorForbiddenNoAuthorizationGroups = (
```
Comment (gunzip) [inline src/responses.ts:461]: this looks unrelated to the PR subject

**Comment by gquadrati** [comment]
> Actually, I'd like the current output more since you still get `value [1] at` which makes no sense if `1` is not in the path anymore.

In this case `value [1]` represents the value passed to `decode` function, so I don't get your point. 
Maybe we could think about changing the value representation between square brackets?


> Moreover the modified output (`root.third_party_data.id`) is ambiguos while `root.1.1.third_party_data.id` is not.

This could be true only internally, but we are returning this representation externally and it's confusing. Users don't care about `t.unions` and `t.intersections` (that causes the indexes).

**Comment by Garma00** [comment]
> Actually, I'd like the current output more

in order to keep the old output maybe we can add a boolean parameter to this function setted `false` as default, then if the caller wants to remove the indexes can pass true to the function.

**Comment by gunzip** [comment]
> > Moreover the modified output (`root.third_party_data.id`) is ambiguos while `root.1.1.third_party_data.id` is not.
> 
> This could be true only internally, but we are returning this representation externally and it's confusing. Users don't care about `t.unions` and `t.intersections` (that causes the indexes).

Maybe I'm wrong, but afaik the numbers are array indexes (isn't it?) that let you rebuild the complete path to the property that carries the wrong value. If is this the case, let's say we have the same property (name) repeated (inside different arrays or nested someway), that would be confusing for the end user. If the numbers represents something else forget about this comment.

**Comment by gquadrati** [comment]
> > > Moreover the modified output (`root.third_party_data.id`) is ambiguos while `root.1.1.third_party_data.id` is not.
> > 
> > 
> > This could be true only internally, but we are returning this representation externally and it's confusing. Users don't care about `t.unions` and `t.intersections` (that causes the indexes).
> 
> Maybe I'm wrong, but afaik the numbers are array indexes (isn't it?) that let you rebuild the complete path to the property that carries the wrong value. If is this the case, let's say we have the same property (name) repeated (inside different arrays or nested someway), that would be confusing for the end user. If the numbers represents something else forget about this comment.

Indexes are caused (in most cases at least) by `t.union` and `t.intersections`, let explain myself with an example that does not contain arrays:

```typescript
const TestType = t.union([
  t.intersection([t.type({ prop1: t.boolean }), t.type({ prop2: t.number })]),
  t.string
]);
```

this is the decoding errors passing `undefined`:

```
value [undefined] at [root.0.0] is not a valid [{ prop1: boolean }]
value [undefined] at [root.0.1] is not a valid [{ prop2: number }]
value [undefined] at [root.1] is not a valid [string]
````

`root.0` and `root.1` are caused by `t.union` position, while `root.0.0` and `root.0.1` are caused by `t.intersection` position.

I'm not sure this is valuable for the end user, I this this is less confusing, with no lack of information:

```
value [undefined] at [root.prop1] is not a valid [{ prop1: boolean }]
value [undefined] at [root.prop2] is not a valid [{ prop2: number }]
value [undefined] at [root] is not a valid [string]
````

Said so, you're right that we need to be careful with arrays.

**Comment by gunzip** [comment]
Thank you, that's my concern (if you strip away all integers you may loose the ability to rebuild the path with arrays).

**Comment by gquadrati** [comment]
@Garma00 will reopen a new PR with some tests, thank you @gunzip
